source/material_model/melt_boukare.cc

```diff
@@ -0,0 +1,1260 @@
+          const double solid_composition = in.composition[q][Fe_solid_idx];
+          double melt_composition = 0.0, melt_molar_fraction = 0.0;
+          double bridgmanite_molar_fraction_in_solid;
+
+          if (this->include_melt_transport())
+            {
+              const unsigned int Fe_melt_idx = this->introspection().compositional_index_for_name("molar_Fe_in_melt");
+              melt_composition = in.composition[q][Fe_melt_idx];
+            }
+
+          convert_composition_to_fraction_of_endmembers(in.temperature[q],
+                                                        solid_composition,
+                                                        melt_composition,
+                                                        endmembers.gibbs_energies,
+                                                        endmember_mole_fractions_per_phase,
+                                                        bridgmanite_molar_fraction_in_solid);
+
+          if (this->include_melt_transport())
+            {
+              const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+
+              melt_molar_fraction = compute_melt_molar_fraction(in.composition[q][porosity_idx],
+                                                                bridgmanite_molar_fraction_in_solid,
+                                                                endmembers,
+                                                                endmember_mole_fractions_per_phase);
+            }
+
+          const double solid_molar_fraction = 1.0 - melt_molar_fraction;
+
+          // Compute endmember molar fractions in the composite.
+          std::vector<double> phase_mole_fractions_in_composite(n_endmembers);
+          double solid_molar_mass = 0.0;
+          double melt_molar_mass = 0.0;
+          double melt_molar_volume = 0.0;
+          double solid_molar_volume = 0.0;
+
+          for (unsigned int i=0; i<n_endmembers; ++i)
+            {
+              // TODO: fix this using a phase vector (phase == 'bridgmanite')
+              const double endmember_phase_fraction_in_solid = i<2 ? bridgmanite_molar_fraction_in_solid : (1.0 - bridgmanite_molar_fraction_in_solid);
+
+              if (endmember_states[i] == EndmemberState::solid)
+                {
+                  phase_mole_fractions_in_composite[i] = solid_molar_fraction * endmember_phase_fraction_in_solid;
+                  endmember_mole_fractions_in_composite[i] = phase_mole_fractions_in_composite[i] * endmember_mole_fractions_per_phase[i];
+                  solid_molar_mass += endmember_mole_fractions_in_composite[i] * molar_masses[i];
+                  solid_molar_volume += endmember_mole_fractions_in_composite[i] * endmembers.volumes[i];
+                }
+              else if (endmember_states[i] == EndmemberState::melt)
+                {
+                  phase_mole_fractions_in_composite[i] = melt_molar_fraction;
```

Comment:
this seems oddly unsymmetrical to line 662 above. Why do we multiply the `solid_molar_fraction` with something, but not the `melt_molar_fraction`?